teps/0104-tasklevel-resource-requirements.md

```diff
@@ -118,63 +118,69 @@ spec:
   resources:
     requests:
       memory: 1Gi
+    limits:
+      memory: 2Gi
 ```
 
 This field should also be added to [PipelineRun.TaskRunSpecs][taskRunSpecs].
 
-### Requests vs Limits
+### Applying Task Resources to Containers
+
+#### Requests
+
+As mentioned in [Resource Requirements in Kubernetes], the effective resource requests of a pod are the sum of the
+resource requests of its containers, and this value is used to determine the resources reserved by the kubelet when scheduling a pod.
+Therefore, when a user configures a resource request for a Task, that request should be applied to one container only in the pod.
+(See [Interaction with LimitRanges](#interaction-with-limitranges) for caveats.)
+All other containers will not have resource requests. This will result in a pod with an effective resource request
+that is the same as that of the Task, and will be scheduled correctly. This is similar to
+Tekton’s [handling of resource requests pre v0.28.0][pre-v28], where the maximum resource request of all containers
+was applied to only one container, and the rest were left without resource requests.
+
+#### Limits
 
 Because Kubernetes considers [containers without resource limits to have higher limits than those with limits configured][effective],
-configuration for limits is more challenging than configuration for requests. Task-level resource limits could be
-implemented several ways, each with their own problems:
+configuration for limits is different than configuration for requests. There are several options for how Task-level resource limits
+could be implemented:
 - If the task resource limit is applied to only one container, the pod will not have an effective limit
 due to the other containers without limits. This defeats the purpose of the feature.
-- If the task limit is applied to each container, the pod has a much higher limit than desired.
-This is especially problematic if requests are not set, because the request will then automatically be set
-to the same value as the limit, and the pod may have difficulty being scheduled.
 - If the task limit is spread out among containers, a task where one step is more resource intensive
 than all the others could get oomkilled or throttled.
-- Limits cannot be dynamically adjusted as steps run, since container resource requirements can't be [updated][update].
+- If the task limit is applied to each container, the pod has a much higher effective limit than desired.
 
-Therefore, this proposal will focus only on task-level resource requests, not limits.
+However, the effective resource limit of a pod are not used for scheduling (see
+[How Pods with resource requests are scheduled][scheduling] and [How Kubernetes applies resource requests and limits][enforcement]).
+Instead, container limits are enforced by the container runtime.
```

Comment:
Question: What happens when the container runtime enforces the limit? Do we currently handle that error somehow? Or does the taskrun just time out?

Reply:
working on the implementation, perhaps can help with some my understanding...

task-level limits will be applied on each step, so each step/container gets enforced with the required limits. Although the task/pod-level will have a larger summed up limits amount(as from each step), the container runtime still follows the limits over the whole process -- due to the sequentially executed steps under Tekton context.

some lines I updated in docs, perhaps help a bit also
https://github.com/tektoncd/pipeline/pull/4877/files?diff=unified&w=0#diff-5007e9db17cf9b70b930577bd627581e10e754347e92e749f5d58614456d7643R82

Reply:
over-limits (task-level) will be checked in a helper func under reconcile and throw an error to fail TaskRun

an example like the current over-limits checks for step-level as
[validateTaskSpecRequestResources()](https://github.com/austinzhao-go/pipeline/blob/tep-0104-per-task-resources-requirements/pkg/reconciler/taskrun/taskrun.go#L337)  
also the [checks implementation](https://github.com/austinzhao-go/pipeline/blob/tep-0104-per-task-resources-requirements/pkg/reconciler/taskrun/validate_resources.go#L226)

Reply:
Discussed in https://github.com/tektoncd/pipeline/issues/4930